### PR TITLE
graph6 implementation - closes #523

### LINKED ISSUE #523: Implement graph6 format
Should be straightforward.

From https://discourse.julialang.org/t/any-julia-package-with-support-for-graph6-graph-encoding/1893

Spec here: http://users.cecs.anu.edu.au/~bdm/data/formats.txt

### PR DESCRIPTION
**Pull request by sbromberger**

### CODE REVIEW

**Comment by sbromberger** [comment]
Paging @jpfairbanks for a review.

**Comment by sbromberger** [comment]
Yeah. I don't know a nice way to get that coverage up.

**Comment by sbromberger** [comment]
@jpfairbanks @vsart review please? I'm afraid I can't squash the commits due to a series of unfortunate merges.

File: src/persistence/graph6.jl [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  l = length(bitstr)
+  padding = k - l
+  bv = falses(k)
+  for i = 1:l
+    bv[padding+i] = (bitstr[i] == '1')
+  end
+  return bv
+end
+
+function R(_x::BitVector)::Vector{UInt8}
```
Comment (jpfairbanks) [inline src/persistence/graph6.jl]: Is this function name from the g6 reference implementation? If not, why the short name? Same goes for the others. Should we call them `g6_R`, `g6_Rp`, etc.?

**Comment by sbromberger** [comment]
`R` and `N` are taken from the specification. `Rp` and `Np` are analogous to `R'` and `N'` (their inverses), which we can't use as valid function names.

File: src/persistence/graph6.jl (hunk 1 above)
Comment (sbromberger) [inline src/persistence/graph6.jl]: I'm ok prefixing them with `g6_`. They are the functions from the spec.

**Comment by sbromberger** [comment]
> We really need to make the persistence testing more thorough and consistent.

Agreed. We should also look into what it would take to integrate with FileIO.jl.

**Comment by vsart** [comment]
I tested `_g6StringToGraph` against some of the graphs from my database and compared these Graphs with the result of the equivalent function I've been using in Python using Sage and they all matched :ok_hand: 

This is really helpful. Thanks!

**Comment by sbromberger** [comment]
@jpfairbanks interesting failure. Random?

**Comment by sbromberger** [comment]
@vsart: try `load(filename,:graph6)` on a file to do the real test. :) Also `save(filename, g, :graph6)` and see if those can be read into python / sage.

**Comment by vsart** [comment]
@sbromberger So I tried to open a file using the `load` function on a file containing 49 graphs but I got a dictionary with 1 entry.

```julia
shell> head -n5 n6_f.g6
E?bo
E?ro
E?qw
E?zo
E?zW

shell> wc n6_f.g6
 49  49 245 n6_f.g6

julia> load("n6_f.g6", :graph6)
Dict{String,LightGraphs.Graph} with 1 entry:
  "g" => {6, 5} undirected graph
```

However, when I opened the file with `f = open('n6_f.g6')` and did subsequent calls to `loadgraph(f, :graph6)` it worked nicely.

Although, trying to load a graph from a file after running out of graphs does give a nasty error. For example:

```julia
julia> f = open("n6_f.g6")
IOStream(<file n6_f.g6>)

julia> for i in 1:49
           g = loadgraph(f, :graph6)
           # Works ok
       end

julia> loadgraph(f, :graph6) # There are no more graphs
ERROR: BoundsError: attempt to access 0-element Array{UInt8,1} at index [1]
 in _g6_Np(::Array{UInt8,1}) at ./REPL[12]:2
 in _g6StringToGraph(::String) at ./REPL[14]:6
 in loadgraph6_mult(::IOStream) at ./REPL[17]:1
 in loadgraph(::IOStream, ::Symbol) at /home/victhor/.julia/v0.5/LightGraphs/src/persistence/common.jl:28
```

Here's a gist with the file: https://gist.github.com/vsart/48bc5bb5aad586ade9e7a3a6c292979e

**Comment by sbromberger** [comment]
Ah, yes. Right now we don't load more than one graph (that is, one line). I can fix this.

**Comment by vsart** [comment]
Tested last commit and it loads all the graphs into a dictionary properly with generic names `gXX` :ok_hand: 

```julia
julia> load("n6_f.g6", :graph6)
Dict{String,LightGraphs.Graph} with 49 entries:
  "g24" => {6, 8} undirected graph
  "g11" => {6, 7} undirected graph
  ⋮     => ⋮
```

**Comment by sbromberger** [comment]
Yeah - I didn't see a way to specify distinguishing names in the graph6 format so I just incremented our standard default.

So I guess it's looking good at this point?

**Comment by vsart** [comment]
LGTM. Thanks again for adding support for this format.

**Comment by sbromberger** [comment]
alrighty - saving multiple graphs is now supported. @vsart, I think this is done: please do one more check of functionality:
1)  load a single graph, 
2)  save a single graph, 
3)  load multiple graphs, 
4)  save multiple graphs (pass in a dict).

@jpfairbanks - please review style / efficiency and the changes I made to SimpleGraphs (basically I had to abstract them since `Dict{String, Graph}` will not match `Dict{String, SimpleGraph}`  - this is also fortunately consistent with what we plan to do in the future anyway).

File: src/persistence/graph6.jl [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  assert(length(x) <= 8 * sizeof(Int))
+  acc = 0
+  for i = 1:length(x)
+    acc = acc << 1 + x[i]
+  end
+  return acc
+end
+
+function _int2bv(n::Int, k::Int)
+  bitstr = lstrip(bits(n), '0')
```
Comment (jpfairbanks) [inline src/persistence/graph6.jl]: This converts n into a String so each number becomes a `64-element Array{UInt8,1}`. Is there a way to interpret an integer directly into a BitVector?

File: src/persistence/graph6.jl [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  l = length(bitstr)
+  padding = k - l
+  bv = falses(k)
```
Comment (jpfairbanks) [inline src/persistence/graph6.jl]: does k need to be an arbitrary integer or just a power of two? We could avoid a lot of overhead using native integer types.

Comment (sbromberger) [inline src/persistence/graph6.jl]: It needs to be a multiple of 18, I believe.

File: src/persistence/graph6.jl (hunk 2 above)
Comment (sbromberger) [inline src/persistence/graph6.jl]: There is no way to do it that I've found. There are probably some more efficient methods but they come at the cost of complexity, and this happens very few times to make it worth much optimization.

File: src/persistence/graph6.jl (hunk 3 above)
Comment (jpfairbanks) [inline src/persistence/graph6.jl]: Can you add a docstring with that info?

Comment (sbromberger) [inline src/persistence/graph6.jl]: I'd rather not encourage docstrings for private methods. :) I can add a comment.

**Comment by jpfairbanks** [comment]
From what I see of the graph6 spec it is for storing dense graphs. I think we can merge this without worrying too much about performance. If someone would like to optimize the implementation, then they can submit a new PR.

**Comment by sbromberger** [comment]
So, I think that now that we have mandatory approvers, it's ok to relax the prohibition on merging your own PRs. Agree? (Just trying to formalize process here.)

**Comment by jpfairbanks** [comment]
Yeah I think that it is ok to merge your own PRs if someone else has approved them.

**Pull request merged**